tempodb/encoding/vparquet/block_traceql.go

```diff
@@ -164,37 +202,86 @@ func operandType(operands traceql.Operands) traceql.StaticType {
 // spansetIterator turns the parquet iterator into the final
 // traceql iterator.  Every row it receives is one spanset.
 type spansetIterator struct {
-	iter parquetquery.Iterator
-}
+	iter   parquetquery.Iterator
+	filter traceql.FilterSpans
 
-var _ traceql.SpansetIterator = (*spansetIterator)(nil)
-
-func (i *spansetIterator) Next(ctx context.Context) (*traceql.Spanset, error) {
+	currentSpans []*traceql.Span
+}
 
-	res, err := i.iter.Next()
-	if err != nil {
-		return nil, err
+func newSpansetIterator(iter parquetquery.Iterator, filter traceql.FilterSpans) *spansetIterator {
+	return &spansetIterator{
+		iter:   iter,
+		filter: filter,
 	}
-	if res == nil {
-		return nil, nil
+}
+
+func (i *spansetIterator) Next() (*traceql.Span, error) {
+	// drain current buffer
+	if len(i.currentSpans) > 0 {
+		ret := i.currentSpans[0]
+		i.currentSpans = i.currentSpans[1:]
+		return ret, nil
 	}
 
-	// The spanset is in the OtherEntries
-	spanset := res.OtherEntries[0].Value.(*traceql.Spanset)
+	for {
+		res, err := i.iter.Next()
+		if err != nil {
+			return nil, err
+		}
+		if res == nil {
+			return nil, nil
+		}
 
-	return spanset, nil
+		// The spanset is in the OtherEntries
+		iface := res.OtherValueFromKey(otherEntrySpansetKey)
+		if iface == nil {
+			return nil, fmt.Errorf("engine assumption broken: spanset not found in other entries")
+		}
+		spanset, ok := iface.(*traceql.Spanset)
+		if !ok {
+			return nil, fmt.Errorf("engine assumption broken: spanset is not of type *traceql.Spanset")
+		}
+
+		// TODO: the engine wants to work with value types, but the fetch layer is using pointers. as a result
+		//  there is some gross "bridge" code here that converts between the two. We should write benchmarks
+		//  and determine if moving to or the other is worth it
+		var filteredSpansets []traceql.Spanset
+		if i.filter != nil {
+			filteredSpansets, err = i.filter(*spanset)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			filteredSpansets = []traceql.Spanset{*spanset}
+		}
+
+		// flatten spans into i.currentSpans
+		for _, ss := range filteredSpansets {
+			for _, s := range ss.Spans {
+				span := s
+				i.currentSpans = append(i.currentSpans, span)
```

Comment:
yes, this is due to the way the metadata portion of the iterators works. this is ok as long as their is only one spanset per trace, but as discussed this will break once we add `by()`.

Reply:
Leaning towards solve when adding `by()` and `coalesce()` so we can benefit from the performance improvements now.